include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+void CellDataStorage<CellIteratorType,DataType>::clear()
+{
+  // Do not call
+  // map.clear();
+  // as we want to be sure no one uses the stored objects. Loop manually:
+  typename std::map<CellIteratorType,CellVector<DataType> >::iterator it = map.begin();
+  while (it != map.end())
+    {
+      // loop over all objects and see if noone is using them
+      for (unsigned int i = 0; i < it->second.size(); i++)
+        {
+          Assert(it->second[i].unique(),
+                 ExcMessage("Can not erase the cell data multiple objects reference its data."));
+        }
+      it = map.erase(it);
+    }
+}
+
+template <typename CellIteratorType, typename DataType>
+template <typename T>
+CellDataStorage<CellIteratorType,DataType>::CellVector<T>
+CellDataStorage<CellIteratorType,DataType>::get_data(const CellIteratorType &cell)
+{
+  static_assert(std::is_base_of<DataType, T>::value,
+                "User's T class should be derived from user's DataType class");
+
+  typename std::map<CellIteratorType,CellVector<DataType> >::iterator it = map.find(cell);
+  Assert(it != map.end(), ExcMessage("Could not find data for the cell"));
+
+  // It would be nice to have a specilized version of this function for T==DataType.
+  // However explicit (i.e full) specialization of a member template is only allowed
+  // when the enclosing class is also explicitly (i.e fully) specialized.
+  // Thus, stick with copying of shared pointers even when the T==DataType:
+  CellVector<T> res (it->second.size());
+  for (unsigned int q = 0; q < res.size(); q++)
+    res[q] = std::dynamic_pointer_cast<T>(it->second[q]);
+  return res;
+}
+
+template <typename CellIteratorType, typename DataType>
+template <typename T>
+CellDataStorage<CellIteratorType,DataType>::CellVector<const T>
+CellDataStorage<CellIteratorType,DataType>::get_data(const CellIteratorType &cell) const
+{
+  static_assert(std::is_base_of<DataType, T>::value,
+                "User's T class should be derived from user's DataType class");
+
+  const typename std::map<CellIteratorType,CellVector<DataType> >::const_iterator it = map.find(cell);
+  Assert(it != map.end(), ExcMessage("Could not find QP data for the cell"));
+
+  // Cast base class to the desired class. This has to be done irrespectively of
+  // T==DataType as we need to return shapred_ptr<const T> to make sure the user
+  // does not modify the content of QP objects
+  CellVector<const T> res (it->second.size());
+  for (unsigned int q = 0; q < res.size(); q++)
+    res[q] = std::dynamic_pointer_cast<const T>(it->second[q]);
+  return res;
+}
+
+//--------------------------------------------------------------------
+//                    ContinuousQuadratureDataTransfer
+//--------------------------------------------------------------------
+
+
+/**
+ * Pack cell data of type @p DataType stored using @p data_storage in @p cell
+ * at each quadrature point to @p matrix_data. Here @p matrix_data is a matrix
```

Comment:
You've already documented this function for doxygen. For regular comments, don't start with `/**` but just a regular `/*`. Same below.